csharp/src/Ice/Reference.cs

```diff
@@ -1825,5 +1555,296 @@ internal void Write(OutputStream ostr)
+
+                    // Wait for the reception of the response.
+                    (response, fin) = await stream.ReceiveResponseFrameAsync(cancel).ConfigureAwait(false);
+
+                    childObserver?.Reply(response.Size);
+
+                    if (!fin)
+                    {
+                        // TODO: handle received stream data.
+                    }
+
+                    // If success, just return the response!
+                    if (response.ResultType == ResultType.Success)
+                    {
+                        return response;
+                    }
+                    observer?.RemoteException();
+                }
+                catch (NoEndpointException ex) when (!cached)
+                {
+                    // If we get NoEndpointException while using non cached endpoints, either all endpoints
+                    // have been excluded or the proxy has no endpoints. we cannot retry, return here to
+                    // preserve any previous exceptions that might have been throw.
+                    childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
+                    observer?.Failed(ex.GetType().FullName ?? "System.Exception"); // TODO cleanup observer logic
+                    return response ?? throw exception ?? ex;
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+                    childObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
+                }
+                finally
+                {
+                    childObserver?.Detach();
+                }
+
+                // Compute retry policy based on the exception or response retry policy, whether or not the connection
+                // is established or the request sent and idempotent
+                Debug.Assert(response != null || exception != null);
+                RetryPolicy retryPolicy =
+                    response?.GetRetryPolicy(this) ?? exception!.GetRetryPolicy(request.IsIdempotent, sent);
+
+                // With the retry-policy OtherReplica we add the current connector to the list of excluded
+                // connectors and remove if from the list of connectors, this prevents the connector to be
+                // tried again during the current retry sequence.
+                if (retryPolicy == RetryPolicy.OtherReplica)
+                {
+                    if ((endpoints?[nextEndpoint] ?? connection?.Endpoint) is Endpoint endpoint)
+                    {
+                        excludedEndpoints ??= new();
+                        excludedEndpoints.Add(endpoint);
+                    }
+                    endpoints?.RemoveAt(nextEndpoint);
+                }
+
+                if (endpoints != null)
+                {
+                    if (connection == null && retryPolicy != RetryPolicy.OtherReplica)
```

Comment:
And change this condition to `if(connection == null || retryPolicy == RetryPolicy.OtherReplica)`. We would move to the next endpoint when connection establishment fails or if the request returned an OtherReplica retry policy exception.